web3/_utils/transactions.py

```diff
@@ -239,3 +256,67 @@ def replace_transaction(
         w3, current_transaction, new_transaction
     )
     return w3.eth.send_transaction(new_transaction)
+
+
+def handle_offchain_lookup(
+    offchain_lookup_payload: Dict[str, Any],
+    transaction: TxParams,
+) -> bytes:
+    formatted_sender = to_hex_if_bytes(offchain_lookup_payload['sender']).lower()
+    formatted_data = to_hex_if_bytes(offchain_lookup_payload['callData']).lower()
+
+    if formatted_sender != to_hex_if_bytes(transaction['to']).lower():
+        raise ValidationError(
+            'Cannot handle OffchainLookup raised inside nested call. Returned `sender` '
+            'value does not equal `to` address in transaction.'
+        )
+
+    for url in offchain_lookup_payload['urls']:
+        formatted_url = URI(
+            str(url)
+            .replace('{sender}', str(formatted_sender))
+            .replace('{data}', str(formatted_data))
+        )
+
+        try:
+            if '{data}' in url and '{sender}' in url:
+                response = get_response_from_get_request(formatted_url)
+            elif '{sender}' in url:
+                response = get_response_from_post_request(formatted_url, data={
+                    "data": formatted_data,
+                    "sender": formatted_sender
+                })
+            else:
+                raise ValidationError('url not formatted properly.')
+        except Exception:
+            continue  # try next url if timeout or issues making the request
+
+        if 400 <= response.status_code <= 499:  # if request returns 400 error, raise exception
+            response.raise_for_status()
+        if not 200 <= response.status_code <= 299:  # if not 400 error, try next url
+            continue
+
+        result = response.json()
+
+        if 'data' not in result.keys():
+            raise ValidationError(
+                "Improperly formatted response for offchain lookup HTTP request - missing 'data' "
+                "field."
+            )
+
+        encoded_data_with_function_selector = b''.join([
+            # 4-byte callback function selector
+            to_bytes_if_hex(offchain_lookup_payload['callbackFunction']),
+
+            # encode the `data` from the result and the `extraData` as bytes
+            encode_abi(
+                ['bytes', 'bytes'],
+                [
+                    to_bytes_if_hex(result['data']),
+                    to_bytes_if_hex(offchain_lookup_payload['extraData']),
+                ]
+            )
+        ])
+
+        return encoded_data_with_function_selector
+    raise Exception("Offchain lookup failed for supplied urls.")
```

Comment:
Do you think a more specific exception here would be useful?

Reply:
I was thinking about that but wasn't sure what would make it better from our arsenal of exceptions. I just raised what I thought to be the equivalent of like a `500` / generic internal server error. Feels like this should be a faulty offchain lookup contract if we ever get here. Open to suggestions.